src/middle/SizedType.ml

```diff
@@ -37,33 +39,55 @@ let collect_exprs st =
 let rec to_unsized = function
   | SInt -> UnsizedType.UInt
   | SReal -> UReal
+  | SComplex -> UComplex
   | SVector _ -> UVector
   | SRowVector _ -> URowVector
   | SMatrix _ -> UMatrix
   | SArray (t, _) -> UArray (to_unsized t)
 
 let rec associate ?init:(assocs = Label.Int_label.Map.empty) = function
-  | SInt | SReal -> assocs
+  | SInt | SReal | SComplex -> assocs
   | SVector (_, e) | SRowVector (_, e) ->
       Expr.Labelled.associate ~init:assocs e
   | SMatrix (_, e1, e2) ->
       Expr.Labelled.(associate ~init:(associate ~init:assocs e1) e2)
   | SArray (st, e) ->
       associate ~init:(Expr.Labelled.associate ~init:assocs e) st
 
-let is_scalar = function SInt | SReal -> true | _ -> false
-let rec inner_type = function SArray (t, _) -> inner_type t | t -> t
+let rec inner_type st = match st with SArray (t, _) -> inner_type t | t -> t
+
+let rec contains_complex st =
+  match st with
+  | SComplex -> true
+  | SArray (t, _) -> contains_complex t
+  | _ -> false
 
 let rec dims_of st =
   match st with
   | SArray (t, _) -> dims_of t
   | SMatrix (_, d1, d2) -> [d1; d2]
   | SRowVector (_, dim) | SVector (_, dim) -> [dim]
+  | SInt | SReal | SComplex -> []
+
+(** Get the dimensions with respect to sizes needed for IO. *)
+let rec get_dims_io st =
+  match st with
   | SInt | SReal -> []
+  | SComplex ->
+      [ Expr.Fixed.
+          { meta=
+              Expr.Typed.Meta.
+                { type_= UnsizedType.UInt
+                ; loc= Location_span.empty
+                ; adlevel= DataOnly }
+          ; pattern= Lit (Int, "2") } ]
+  | SVector (_, d) | SRowVector (_, d) -> [d]
+  | SMatrix (_, dim1, dim2) -> [dim1; dim2]
+  | SArray (t, dim) -> dim :: get_dims_io t
 
 let rec get_dims st =
   match st with
-  | SInt | SReal -> []
+  | SInt | SReal | SComplex -> []
   | SVector (_, d) | SRowVector (_, d) -> [d]
   | SMatrix (_, dim1, dim2) -> [dim1; dim2]
   | SArray (t, dim) -> dim :: get_dims t
```

Comment:
A small note here, we essentially need two ways to query dimensions. In one case we treat the complex type as a dual number when we are talking to the I/O to tell it how many parameters we have. In the other we think about complex types from the users perspective where we treat the complex type as a scalar. To differentiate between the two I just made a `get_dims_io` and a `get_dims` as seperate methods. Is that cool?

Reply:
I think that’s a good approach (cleaner than what I had done), I would just add to the comment calling out that the primary difference is in Complex types so it’s clear at a glance. 

Also, it’s nit picky but can we use Expr.Helpers.int in the _io version? 

Otherwise LGTM!

Reply:
Yes I like both of those!